File: python/arikernel/runtime/hash_chain.py
"""SHA-256 hash chain for tamper-evident audit logging.

Compatible with the TypeScript runtime's hash chain format.
"""

from __future__ import annotations

import hashlib

GENESIS_HASH = "0" * 64


def compute_hash(data: str, previous_hash: str) -> str:
    """Compute SHA-256 hash: update(previousHash) then update(data).

    This matches the TypeScript implementation exactly:
        createHash('sha256').update(previousHash).update(data).digest('hex')
    """
    h = hashlib.sha256()
    h.update(previous_hash.encode("utf-8"))
    h.update(data.encode("utf-8"))
    return h.hexdigest()


def verify_chain(
    events: list[dict],
) -> dict:
    """Verify a chain of events.

    Args:
        events: List of dicts with 'hash', 'previousHash', 'data' keys.

    Returns:
        {"valid": True} or {"valid": False, "brokenAt": index}
    """
    for i, event in enumerate(events):
        expected = compute_hash(event["data"], event["previousHash"])
        if expected != event["hash"]:
            return {"valid": False, "brokenAt": i}
        if i > 0 and event["previousHash"] != events[i - 1]["hash"]:
            return {"valid": False, "brokenAt": i}
    return {"valid": True}

File: packages/audit-log/src/hash-chain.ts
/**
 * SHA-256 hash chain for tamper-evident audit logging.
 *
 * Provides local tamper evidence: any modification, deletion, or insertion
 * of events breaks the chain. When an HMAC key is provided, each hash is
 * an HMAC-SHA256 keyed digest — an attacker with filesystem access cannot
 * recompute valid hashes without the key.
 *
 * For production deployments, forward events to an external append-only
 * store (SIEM, CloudTrail, immutable log) for defense-in-depth.
 */
import { createHash, createHmac } from "node:crypto";

const GENESIS_HASH = "0".repeat(64);

/**
 * Compute a chain hash over (previousHash || data).
 * If an HMAC key is provided, uses HMAC-SHA256 so an attacker cannot
 * forge hashes without the key — even with full SQLite access.
 */
export function computeHash(data: string, previousHash: string, hmacKey?: Buffer): string {
	if (hmacKey) {
		return createHmac("sha256", hmacKey).update(previousHash).update(data).digest("hex");
	}
	return createHash("sha256").update(previousHash).update(data).digest("hex");
}

export function genesisHash(): string {
	return GENESIS_HASH;
}

export function verifyChain(
	events: Array<{ hash: string; previousHash: string; data: string }>,
	hmacKey?: Buffer,
): {
	valid: boolean;
	brokenAt?: number;
} {
	for (let i = 0; i < events.length; i++) {
		const event = events[i];
		const expected = computeHash(event.data, event.previousHash, hmacKey);
		if (expected !== event.hash) {
			return { valid: false, brokenAt: i };
		}
		if (i > 0 && event.previousHash !== events[i - 1].hash) {
			return { valid: false, brokenAt: i };
		}
	}
	return { valid: true };
}
